src/Symfony/Component/Config/Resource/ReflectionClassResource.php

```diff
@@ -0,0 +1,185 @@
+    {
+        $this->className = $classReflector->name;
+        $this->classReflector = $classReflector;
+    }
+
+    public function isFresh($timestamp)
+    {
+        if (null === $this->hash) {
+            $this->hash = $this->computeHash();
+            $this->loadFiles($this->classReflector);
+        }
+
+        foreach ($this->files as $file => $v) {
+            if (!file_exists($file)) {
+                return false;
+            }
+
+            if (null === $mtime = &self::$mtimeCache[$file]) {
+                $mtime = @filemtime($file);
+            }
+            if ($mtime > $timestamp) {
+                if ($this->hash !== $this->computeHash()) {
+                    return false;
+                }
+
+                // ensure the next checks won't require a hash computation again,
+                // use a local mtime cache so that the current process still sees
+                // the original mtime and validates the hash for child classes
+                @touch($file, $timestamp);
+
+                return true;
+            }
+        }
+
+        return true;
+    }
+
+    public function __toString()
+    {
+        return 'reflection.'.$this->className;
+    }
+
+    public function serialize()
+    {
+        if (null === $this->hash) {
+            $this->hash = $this->computeHash();
+            $this->loadFiles($this->classReflector);
+        }
+
+        return serialize(array($this->files, $this->className, $this->hash));
+    }
+
+    public function unserialize($serialized)
+    {
+        list($this->files, $this->className, $this->hash) = unserialize($serialized);
+    }
+
+    private function loadFiles(\ReflectionClass $class)
+    {
+        foreach ($class->getInterfaces() as $v) {
+            $this->loadFiles($v);
+        }
+        do {
+            $file = $class->getFileName();
+            if (false !== $file && file_exists($file)) {
+                $this->files[$file] = null;
+            }
+            foreach ($class->getTraits() as $v) {
+                $this->loadFiles($v);
+            }
+        } while ($class = $class->getParentClass());
+    }
+
+    private function computeHash()
+    {
+        if (null === $this->classReflector) {
+            try {
+                $this->classReflector = new \ReflectionClass($this->className);
+            } catch (\ReflectionException $e) {
+                // the class does not exist anymore
+                return false;
+            }
+        }
+        $hash = hash_init('md5');
+
+        foreach ($this->generateSignature($this->classReflector) as $info) {
+            hash_update($hash, $info);
+        }
+
+        return hash_final($hash);
+    }
+
+    private function generateSignature(\ReflectionClass $class)
+    {
+        yield $class->getDocComment().$class->getModifiers();
+
+        if ($class->isTrait()) {
+            yield print_r(class_uses($class->name), true);
```

Comment:
no need: traits are not directly part of the signature - the method they provide will be seen as any other methods, thus tracked